### PR TITLE
camera: adding subscription APIs

### PR DESCRIPTION
**Pull request by julianoes**
This PR adds missing camera subscription APIs:
- [x] camera mode
- [x] video stream info
- [x] capture info
- [x] camera status
- [x] current setting
- [x] possible settings

Connects to #334.

### CODE REVIEW

File: plugins/camera/camera_impl.cpp [hunk 1]
```diff
@@ -410,6 +413,14 @@ Camera::Result CameraImpl::get_video_stream_info(Camera::VideoStreamInfo &info)
     ;
 }
 
+void CameraImpl::subscribe_video_stream_info(
+    const Camera::subscribe_video_stream_info_callback_t callback)
+{
+    // FIXME: This will only work if you also send a request to get the video stream info,
```
[the remark this answers is not in the document]
Comment (julianoes) [reply plugins/camera/camera_impl.cpp]: Correct, we just have to do a `call_every`, not that hard :smile:

[the remark this answers is not in the document]
Comment (julianoes) [reply plugins/camera/camera_impl.cpp]: I just haven't ever used the streaming API and there are no tests, so I didn't bother.

**Comment by julianoes** [comment]
> Is it right to say that the camera mode is advertised by the camera

I think right now it is not advertised as such. It's just when you set it you also receive the feedback and use that. We potentially could add polling the mode as well in order to see changes made by e.g. QGroundControl.

File: plugins/camera/camera_impl.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (julianoes) [reply plugins/camera/camera_impl.cpp]: Done but untested.

**Comment by JonasVautherin** [comment]
> We potentially could add polling the mode as well in order to see changes made by e.g. QGroundControl.

But if QGC sets the mode, won't the camera advertise it? Anyway it's fine for now.

**Comment by julianoes** [comment]
> But if QGC sets the mode, won't the camera advertise it? Anyway it's fine for now.

It could/should advertise it using: http://mavlink.org/messages/common#CAMERA_SETTINGS. We probably need to add that to E90.

**Comment by julianoes** [comment]
@JonasVautherin this should be complete but I'm not happy with the naming yet.

Currently it's all `options` but we need to distinguish between a "selected option" (a setting) and a "selectable option" (an option).

**Comment by julianoes** [comment]
Currently we have:
```
typedef std::function<void(const std::vector<std::pair<std::string, std::string>>)>
    subscribe_current_options_callback_t;

typedef std::function<void(const std::vector<std::pair<std::string, std::vector<std::string>>>)>
    subscribe_possible_options_callback_t;
```

We could refactor that to:
```
struct Option {
    std::string id;
};

struct Setting {
    id;
    Option option;
};

struct SettingList {
    id;
    std::vector<Option> options;
};

typedef std::function<void(const std::vector<Setting>)>
    subscribe_current_settings_callback_t;

typedef std::function<void(const std::vector<SettingList>)>
    subscribe_possible_settings_callback_t;
```

@JonasVautherin what do you think?

**Comment by JonasVautherin** [comment]
I like your proposition. I would suggest `SettingOptions` instead of `SettingList`, though :-).

**Comment by julianoes** [comment]
Ok, I'll refactor it.

**Pull request merged**